File: advent2023/05.zig

const std = @import("std");
const input = @embedFile("input05.txt");

const Range = struct {
    i: u64,
    j: u64,
};

pub fn main() !void {
    var buf: [4000]u8 = undefined;
    var fba = std.heap.FixedBufferAllocator.init(&buf);
    const alloc = fba.allocator();
    const w = std.io.getStdOut().writer();

    var maps = std.mem.splitSequence(u8, input, "\n\n");
    const seedstr = maps.next().?;
    var seeds_it = std.mem.splitScalar(u8, seedstr, ' ');
    _ = seeds_it.next(); // drop "seeds:"
    var seeds = try std.ArrayList(struct{s: u64, moved:bool}).initCapacity(alloc, 30);
    var moved = try std.ArrayList(Range).initCapacity(alloc, 100);
    var unmoved = try std.ArrayList(Range).initCapacity(alloc, 100);
    while (seeds_it.next()) |seed_start| {
        const seed_len = seeds_it.next().?;
        const i = try std.fmt.parseInt(u64, seed_start, 10);
        const len = try std.fmt.parseInt(u64, seed_len, 10);
        try unmoved.append(Range{
            .i = i,
            .j = i + len,
        });
        try seeds.append(.{.s=i, .moved=false});
        try seeds.append(.{.s=len, .moved=false});
    }
    // move the ranges and original seeds according to maps
    while (maps.next()) |map| {
        var lines = std.mem.splitScalar(u8, map, '\n');
        _ = lines.next(); // drop description
        while (lines.next()) |map_part| {
            if (map_part.len==0) continue;
            var n_it = std.mem.splitScalar(u8, map_part, ' ');
            const dst = try std.fmt.parseInt(u64, n_it.next().?, 10);
            const src = try std.fmt.parseInt(u64, n_it.next().?, 10);
            const len = try std.fmt.parseInt(u64, n_it.next().?, 10);
            const end = src + len;
            // move seeds according to map part
            for (seeds.items) |*seed| {
                if (seed.moved) continue;
                if (src <= seed.s and seed.s < end) {
                    seed.s +%= dst-%src;
                    seed.moved = true;
                }
            }
            var i: usize = 0;
            while (i < unmoved.items.len) {
                // intersect all ranges with the map part
                const r = unmoved.items[i];
                if (src < r.j and r.i < end) {
                    _ = unmoved.swapRemove(i);
                    const new_r = Range {.i= @max(r.i, src), .j = @min(r.j, end)};
                    if (r.i < new_r.i) {
                        try unmoved.append(.{.i = r.i, .j = new_r.i});
                    }
                    if (new_r.j < r.j) {
                        try unmoved.append(.{.i = new_r.j, .j = r.j});
                    }
                    try moved.append(.{.i = new_r.i - src+dst, .j = new_r.j - src+dst});
                } else {
                    i += 1;
                }
            }
        }
        // finished with this section; collect moved ranges
        try unmoved.appendSlice(moved.items);
        moved.clearRetainingCapacity();
        for (seeds.items) |*seed| {
            seed.moved = false;
        }
    }
    var min_start: u64 = unmoved.items[0].i;
    for (unmoved.items[1..]) |it| {
        if (it.i < min_start) min_start = it.i;
    }
    var min_seed: u64 = seeds.items[0].s;
    for (seeds.items[1..]) |seed| {
        if (seed.s < min_seed) min_seed = seed.s;
    }
    try w.print("part1: {}\npart2: {}\n", .{min_seed, min_start});
}

File: advent2023/05.rs

use std::cmp::{max,min};
use itertools::Itertools;

const INPUT: &str = include_str!("input05.txt");

struct Seed {
    s: i64,
    moved: bool,
}
struct Range {
    i: i64,
    j: i64,
    moved: bool,
}

fn main() {
    let mut maps = INPUT.split("\n\n");
    let seedstr = maps.next().unwrap();
    let iseeds: Vec<i64> = seedstr.split(' ').skip(1).map(|x| x.parse::<i64>().unwrap()).collect();
    let mut ranges: Vec<Range> = iseeds.iter().tuples().map(|(s,len)| Range{i: *s, j: *s+*len, moved: false}).collect();
    let mut seeds: Vec<Seed> = iseeds.into_iter().map(|x| Seed{s: x, moved: false}).collect();
    // move the ranges and original seeds according to maps
    for map in maps {
        for map_part in map.lines().skip(1) {
            let (dst, src, len) = map_part.split(' ').map(|x| x.parse::<i64>().unwrap()).next_tuple().unwrap();
            let end = src + len;
            // move seeds according to map part
            for seed in &mut seeds {
                if seed.moved {continue;}
                if src <= seed.s && seed.s < end {
                    seed.s += dst-src;
                    seed.moved = true;
                }
            }
            for i in 0..ranges.len() {
                // intersect all ranges with the map part
                //let r = &ranges[i];
                if ranges[i].moved {continue;}
                let (ri, rj) = (ranges[i].i, ranges[i].j);
                if src < rj && ri < end {
                    let new_i = max(ri, src);
                    let new_j = min(rj, end);
                    if ri < new_i {
                        ranges.push(Range {i: ri, j: new_i, moved: false});
                    }
                    if new_j < rj {
                        ranges.push(Range {i: new_j, j: rj, moved: false});
                    }
                    ranges[i] = Range {i: new_i - src+dst, j: new_j - src+dst, moved: true};
                }
            }
        }
        // finished with this section; collect moved ranges
        for seed in &mut seeds {
            seed.moved = false;
        }
        for range in &mut ranges {
            range.moved = false;
        }
    }
    let part1 = seeds.into_iter().map(|x| x.s).min().unwrap();
    let part2 = ranges.into_iter().map(|x| x.i).min().unwrap();
    println!("part1: {}\npart2: {}\n", part1, part2);
}
